samples/ipc/openamp: Introduce OpenAMP v2018.10 Remote Echo Sample

diegosueiro opened the pull request:
Introduces the OpenAMP Remote Echo Sample to have Linux and Zephyr communicating in i.MX7/6SX SoCs.

Note that this PR depends on #11355.

galak committed:
samples: openamp: Update sample to OpenAMP v2018.10 release

The v2018.10 release of OpenAMP reworks the API set and splits the
remoteproc vs rpmsg interfaces so one can use rpmsg without remoteproc.

This helps drastically reduce the code footprint utilized by OpenAMP.
The remote see around 4k reduction in code size.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
ext: libmetal: Update libmetal to v2018.10 release

Since there's an official release lets update to it.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
ext: open-amp: Update OpenAMP to v2018.10 release

Update to new OpenAMP v2018.10 release.  This release allows us to
utilize just rpmsg without remoteproc.  The API set has changed and
requires updates to the openamp sample.  Additionally, the changes
in this release reduce the code size footprint, and support a static
allocation memory model.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

diegosueiro committed:
samples/ipc/openamp: Introduce OpenAMP Remote Echo Sample

Introduces the OpenAMP Remote Echo Sample to have Linux and Zephyr
communicating in i.MX7/6SX SoCs.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro commented:
ping @galak and @wjliang.

diegosueiro commented:
Just need to review the last commit: `samples/ipc/openamp: Introduce OpenAMP Remote Echo Sample`

diegosueiro commented:
> @diegosueiro : did you merge it? you shall rebase it.

I rebased it. I'm not sure what happened.

This was the way I created this branch:
`git fetch upstream  pull/11355/head:openamp-v2018.10-imx-sample`

Then I realized that I had to update my local branch (openamp-v2018.10-imx-sample) based on PR 11355. To do that I did:

```
git remote add galak https://github.com/galak/zephyr.git
git fetch galak
git checkout openamp-v2018.10-imx-sample
git rebase galak/openamp-v2018.10
git push --force origin openamp-v2018.10-imx-sample
```

And this created all this mess.

I'm really sorry guys.
I don't know if there is a way of cleaning up this or if it is better to close this PR and create another one.

jhedberg commented:
@diegosueiro since you're submitting the pull request for the upstream master branch, that's what you need to rebase against. You can always check prior to pushing what commits will be included using e.g. "git log origin/master.." (assuming "origin" is the upstream tree).

diegosueiro commented:
> @diegosueiro since you're submitting the pull request for the upstream master branch, that's what you need to rebase against. You can always check prior to pushing what commits will be included using e.g. "git log origin/master.." (assuming "origin" is the upstream tree).

Understood. Let me try it again.

diegosueiro commented:
> > @diegosueiro since you're submitting the pull request for the upstream master branch, that's what you need to rebase against. You can always check prior to pushing what commits will be included using e.g. "git log origin/master.." (assuming "origin" is the upstream tree).
> 
> Understood. Let me try it again.

@jhedberg,
It seems that it worked.

@all,

Sorry about the mess again.

diegosueiro commented:
@stanislav-poboril,

Can you please test this sample on UDOO Neo Full?

stanislav-poboril commented on samples/subsys/ipc/openamp/remote_echo/boards/udoo_neo_full.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_HEAP_MEM_POOL_SIZE=16384
+CONFIG_IPM_IMX=y
+CONFIG_IPM_IMX_MAX_DATA_SIZE_4=y
```
Could you please rename the file to udoo_neo_full_m4.conf?

diegosueiro replied:
addressed

stanislav-poboril commented:
> Can you please test this sample on UDOO Neo Full?

I've tried, there are two changes needed that I described in a review.
But I still was not successful yet. I will try to look at it on Wednesday.

diegosueiro commented:
> > Can you please test this sample on UDOO Neo Full?
> 
> I've tried, there are two changes needed that I described in a review.
> But I still was not successful yet. I will try to look at it on Wednesday.

Did you have anything printed out on Zephyr console?
What kernel Linux source are you using?
Did you check the devicetree nodes in the Linux side that I put in the README.rst?
What is the output of `dmesg | grep -iE "virtio|vring|rpmsg|virtqueue"` in the Linux?

diegosueiro commented on samples/subsys/ipc/openamp/remote_echo/common.h:
```diff
@@ -0,0 +1,31 @@
+
+#define RPMSG_CHAN_NAME         "rpmsg-openamp-demo-channel"
+
+#if defined(CONFIG_SOC_SERIES_IMX7_M4) || defined(CONFIG_SOC_SERIES_IMX6_M4)
```
adressed

dbkinder commented on samples/subsys/ipc/openamp/remote_echo/README.rst:
```diff
@@ -0,0 +1,144 @@
+
+Open a serial terminal (minicom, putty, etc.) and connect the board with the
+following settings in the UART_2 for colibri_imx7d_m4 and warp7_m4 boards and
+UART_5 for udoo_neo_full_m4 board:
```
```suggestion
UART_5 for the udoo_neo_full_m4 board:
```

dbkinder commented on samples/subsys/ipc/openamp/remote_echo/README.rst:
```diff
@@ -0,0 +1,144 @@
+- Parity: None
+- Stop bits: 1
+
+Below you will find the instructions to load and run Zephyr on M4 from A7 using
```
```suggestion
Here are the instructions to load and run Zephyr on M4 from A7 using
```

dbkinder commented on samples/subsys/ipc/openamp/remote_echo/README.rst:
```diff
@@ -0,0 +1,144 @@
+Below you will find the instructions to load and run Zephyr on M4 from A7 using
+u-boot.
+
+Copy the compiled zephyr.bin to the first EXT partition of the SD card and plug
```
```suggestion
Copy the compiled ``zephyr.bin`` to the first EXT partition of the SD card and plug
```

dbkinder commented on samples/subsys/ipc/openamp/remote_echo/README.rst:
```diff
@@ -0,0 +1,144 @@
+u-boot.
+
+Copy the compiled zephyr.bin to the first EXT partition of the SD card and plug
+into the board. Power it up and stop the u-boot execution.
```
```suggestion
the SD card into the board. Power up the board and stop the u-boot execution.
```

dbkinder commented on samples/subsys/ipc/openamp/remote_echo/README.rst:
```diff
@@ -0,0 +1,144 @@
+
+Copy the compiled zephyr.bin to the first EXT partition of the SD card and plug
+into the board. Power it up and stop the u-boot execution.
+Set the u-boot environment variables and run the zephyr.bin from the TCML
```
```suggestion
Set the u-boot environment variables and run the ``zephyr.bin`` from the TCML
```

dbkinder commented on samples/subsys/ipc/openamp/remote_echo/README.rst:
```diff
@@ -0,0 +1,144 @@
+Copy the compiled zephyr.bin to the first EXT partition of the SD card and plug
+into the board. Power it up and stop the u-boot execution.
+Set the u-boot environment variables and run the zephyr.bin from the TCML
+memory.
```
```suggestion
memory, as described here:
```

stanislav-poboril commented:
@diegosueiro 

> Did you have anything printed out on Zephyr console?

If I remember it correctly I've got -2005 on endpoint creation on Zephyr console. Maybe just some memory allocation problem or something on Zephyr side.

> What kernel Linux source are you using?

I've been using UDOObuntu version of Linux kernel (something old like 3.15), patched for RPMsg - it worked with RPMsg-Lite examples previously.

> Did you check the devicetree nodes in the Linux side that I put in the README.rst?

Yes, it matches (except for the shared memory region which I have adjusted correspondingly in the Zephyr application).

> What is the output of `dmesg | grep -iE "virtio|vring|rpmsg|virtqueue"` in the Linux?

I will try it on Wednesday. But I remember grep -i rpmsg returned just two rows. If I modprobe imx_rpmsg_tty, it proceeds without error but the /dev/ttyRPMSG is not created.

diegosueiro commented:
@stanislav-poboril ,

> I've been using UDOObuntu version of Linux kernel (something old like 3.15), patched for RPMsg - it worked with RPMsg-Lite examples previously.

Can you please make this change on this [file](https://github.com/UDOOboard/linux_kernel/blob/3.14-1.0.x-udoo/drivers/rpmsg/virtio_rpmsg_bus.c#L980) and post the output?

```
-       dev_dbg(&vdev->dev, "buffers: va %p, dma 0x%llx\n", bufs_va,
+       dev_info(&vdev->dev, "buffers: va %p, dma 0x%llx\n", bufs_va,
```

diegosueiro commented:
@nashif and @galak,

I have no idea how to fix this sanitycheck error.

stanislav-poboril commented:
@diegosueiro 

> dmesg | grep -iE "virtio|vring|rpmsg|virtqueue

    # dmesg | grep -iE "virtio|vring|rpmsg|virtqueue"
    [    0.136597] virtio_rpmsg_bus virtio0: rpmsg host is online
    [    0.136695] imx rpmsg driver is registered.

> Can you please make this change on this file and post the output?

Sorry I could not test this change in the Linux kernel today. And I am not available till end of the year. Maybe we can remove this example from UDOO board?

This return 0.

    uint32_t virtqueue_get_desc_size(struct virtqueue *vq)
    {
	uint16_t head_idx = 0;
	uint16_t avail_idx = 0;
	uint32_t len = 0;
	printk("vq->vq_available_idx %u\n", vq->vq_available_idx);
	printk("vq->vq_ring.avail->idx %u\n", vq->vq_ring.avail->idx);
	if (vq->vq_available_idx == vq->vq_ring.avail->idx) {
		printk("ret 0\n");
		return 0;
	}


        ***** Booting Zephyr OS zephyr-v1.13.0-2509-g1daf559 *****
    Starting application thread!
    RPMSG_BUFFER_SIZE = 512
    virtqueue_allocate sizeof(struct virtqueue)=56, num_desc_extra=256, sizeof(struct vq_desc_extra)=8
    vq_size 2104
    virtqueue_allocate sizeof(struct virtqueue)=56, num_desc_extra=256, sizeof(struct vq_desc_extra)=8
    vq_size 2104
    vq->vq_available_idx 0
    vq->vq_ring.avail->idx 0
    ret 0
    vq->vq_available_idx 0
    vq->vq_ring.avail->idx 0
    ret 0
    desc size 0
    vq->vq_available_idx 0
    vq->vq_ring.avail->idx 0
    ret 0
    desc size u 0
    sizeof(struct rpmsg_hdr) 16
    src 0
    dst 53
    avail_size -16
    size 40
    rpmsg_create_ept failed: -2005
    OpenAMP remote echo demo ended.

diegosueiro commented:
> @diegosueiro
> 
> > dmesg | grep -iE "virtio|vring|rpmsg|virtqueue
> 
> ```
> # dmesg | grep -iE "virtio|vring|rpmsg|virtqueue"
> [    0.136597] virtio_rpmsg_bus virtio0: rpmsg host is online
> [    0.136695] imx rpmsg driver is registered.
> ```
> > Can you please make this change on this file and post the output?
> 
> Sorry I could not test this change in the Linux kernel today. And I am not available till end of the year. Maybe we can remove this example from UDOO board?
> 
This print will help to see what is the beginning of the shared memory address.
No problem. I can remove the i.MX6SX from this sample.

> This return 0.
> 
> ```
> uint32_t virtqueue_get_desc_size(struct virtqueue *vq)
> {
> uint16_t head_idx = 0;
> uint16_t avail_idx = 0;
> uint32_t len = 0;
> printk("vq->vq_available_idx %u\n", vq->vq_available_idx);
> printk("vq->vq_ring.avail->idx %u\n", vq->vq_ring.avail->idx);
> if (vq->vq_available_idx == vq->vq_ring.avail->idx) {
> 	printk("ret 0\n");
> 	return 0;
> }
> 
> 
>     ***** Booting Zephyr OS zephyr-v1.13.0-2509-g1daf559 *****
> Starting application thread!
> RPMSG_BUFFER_SIZE = 512
> virtqueue_allocate sizeof(struct virtqueue)=56, num_desc_extra=256, sizeof(struct vq_desc_extra)=8
> vq_size 2104
> virtqueue_allocate sizeof(struct virtqueue)=56, num_desc_extra=256, sizeof(struct vq_desc_extra)=8
> vq_size 2104
> vq->vq_available_idx 0
> vq->vq_ring.avail->idx 0
> ret 0
> vq->vq_available_idx 0
> vq->vq_ring.avail->idx 0
> ret 0
> desc size 0
> vq->vq_available_idx 0
> vq->vq_ring.avail->idx 0
> ret 0
> desc size u 0
> sizeof(struct rpmsg_hdr) 16
> src 0
> dst 53
> avail_size -16
> size 40
> rpmsg_create_ept failed: -2005
> OpenAMP remote echo demo ended.
> ```
This can be happening because it is looking into the wrong VRING_TX address. In the VRING_TX the master puts the available virtqueues. I'm not sure if it is possible to plug a debugger into the Udoo Neo, because using a debugger helped me a lot.

galak committed:
boards: lpcxpresso54114: lpcxpresso54114_m0 enable SEGGER support by default

The lpcxpresso54114_m0 tends to be the secondary core so doesn't
typically have a UART setup for it.  We had SEGGER support enabled by
default previously, but the new Kconfig option CONFIG_USE_SEGGER_RTT
needs to be set as well now.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

diegosueiro commented:
@stanislav-poboril,
> > Sorry I could not test this change in the Linux kernel today. And I am not available till end of the year. Maybe we can remove this example from UDOO board?
> 
> This print will help to see what is the beginning of the shared memory address.
> No problem. I can remove the i.MX6SX from this sample.
> 
> >

I'll wait for the others review because I think this PR will be merged next year anyway. And maybe we will have time to debug this on the Udoo Neo.

Do you have a public repo where I can see the imx_rpmsg driver ported to the Udoo Neo kernel?

diegosueiro commented:
@arnop2 and @wjliang, 

Can you please review this PR?

arnop2 commented:
@diegosueiro
i had a look to your PR.  First of all, I would like to understand the context ( not familiar with the IMX environment). It is not the way i would implement the example, but probably it is just not the same context as mine...
- Your sample concerns inter-processor  communication between a Linux and a Zephyr, right?
- On Linux side do you use the generic remoteproc framework to load the Zephyr firmware and the generic rpmsg framework for the IPC? Because i can not see any resource table that should be needed in this case.
I sent to Kumar few month ago a sample to echo the Linux rpmsg sample client. Don't know if this could help you...it is based on resource table usage.

diegosueiro commented:
@arnop2,

Thanks for taking time to look into this.

> @diegosueiro
> i had a look to your PR. First of all, I would like to understand the context ( not familiar with the IMX environment). It is not the way i would implement the example, but probably it is just not the same context as mine...
> 
> * Your sample concerns inter-processor  communication between a Linux and a Zephyr, right?

Yes, that is right.

> * On Linux side do you use the generic remoteproc framework to load the Zephyr firmware and the generic rpmsg framework for the IPC? Because i can not see any resource table that should be needed in this case.

For the i.MX processors in the mainline there are only the mailbox driver and the remoteproc, but the later doesn't support "extracting" the resource table data from the remote firmware binary. And there is no rpmsg driver.
But, NXP provides in their source repository the rpmsg driver where the vrings data comes from the device tree. You can use [this repo](https://github.com/Freescale/linux-fslc/blob/4.9-1.0.x-imx/drivers/rpmsg/imx_rpmsg.c) as a reference.
The remote firmware loading can be done by u-boot or a [userspace tool](https://github.com/NXPmicro/imx-m4fwloader).

This why we don't need to use the remoteproc framework for the lifecycle management and the vring data extraction.

I already have Zephyr and Linux communicating by using the RPMsg-lite: [slides](https://events.linuxfoundation.org/wp-content/uploads/2017/12/Linux-and-Zephyr-%E2%80%9CTalking%E2%80%9D-to-Each-Other-in-the-Same-SoC-Diego-Sueiro-Sepura-Embarcados-1.pdf) and [video](https://youtu.be/4zMyjaEgj2U).


>   I sent to Kumar few month ago a sample to echo the Linux rpmsg sample client. Don't know if this could help you...it is based on resource table usage.

arnop2 commented:
> @arnop2,
> 
> Thanks for taking time to look into this.
> 
> > @diegosueiro
> > i had a look to your PR. First of all, I would like to understand the context ( not familiar with the IMX environment). It is not the way i would implement the example, but probably it is just not the same context as mine...
> > 
> > * Your sample concerns inter-processor  communication between a Linux and a Zephyr, right?
> 
> Yes, that is right.
> 
> > * On Linux side do you use the generic remoteproc framework to load the Zephyr firmware and the generic rpmsg framework for the IPC? Because i can not see any resource table that should be needed in this case.
> 
> For the i.MX processors in the mainline there are only the mailbox driver and the remoteproc, but the later doesn't support "extracting" the resource table data from the remote firmware binary. And there is no rpmsg driver.
> But, NXP provides in their source repository the rpmsg driver where the vrings data comes from the device tree. You can use [this repo](https://github.com/Freescale/linux-fslc/blob/4.9-1.0.x-imx/drivers/rpmsg/imx_rpmsg.c) as a reference.
> The remote firmware loading can be done by u-boot or a [userspace tool](https://github.com/NXPmicro/imx-m4fwloader).
> 
> This why we don't need to use the remoteproc framework for the lifecycle management and the vring data extraction.
> 
> I already have Zephyr and Linux communicating by using the RPMsg-lite: [slides](https://events.linuxfoundation.org/wp-content/uploads/2017/12/Linux-and-Zephyr-%E2%80%9CTalking%E2%80%9D-to-Each-Other-in-the-Same-SoC-Diego-Sueiro-Sepura-Embarcados-1.pdf) and [video](https://youtu.be/4zMyjaEgj2U).

Thank for the clarification! now i better understand your implementation.

diegosueiro commented:
> @diegosueiro
> 
> I reviewed your patch, i did not see anything wrong.
> But my concern here is that it is not the generic way to implement communication with the Linux OS, but IMX platform specific. From my point of view a good example would be to answer to the rpmsg client sample driver. This drivers follow the generic implementation of the rpmsg protocol in Linux, so that would allows Zephyr user to simply test it without implementing a specific Linux drivers and/or a specific zephyr application. Do you plan also to implement it?

At the moment this sample is tight up to the iMX processors and RPMsg driver implementation in the Linux side only for the virtio operations. Is there any other place in this sample that is specific to the imx? 
I'm not planning to upstream a more generic Linux RPMsg driver in the short term.

arnop2 commented:
> > @diegosueiro
> > I reviewed your patch, i did not see anything wrong.
> > But my concern here is that it is not the generic way to implement communication with the Linux OS, but IMX platform specific. From my point of view a good example would be to answer to the rpmsg client sample driver. This drivers follow the generic implementation of the rpmsg protocol in Linux, so that would allows Zephyr user to simply test it without implementing a specific Linux drivers and/or a specific zephyr application. Do you plan also to implement it?
> 
> At the moment this sample is tight up to the iMX processors and RPMsg driver implementation in the Linux side only for the virtio operations. Is there any other place in this sample that is specific to the imx?
> I'm not planning to upstream a more generic Linux RPMsg driver in the short term.

i'm not sure that we speak about the same things, so lets me precise my thinking. When i spoke about the rpmsg sample client, i was referred to this one that is already in the kernel mainline 
https://elixir.bootlin.com/linux/latest/source/samples/rpmsg/rpmsg_client_sample.c
what i would expect is that the sample echo answer to this driver. 
your sample is compatible with the NXP branch based on a Kernel 4.9, but i 'm not sure that it is straightforward to integrate it on another platform with the lasted Kernel. 
This is my concern. So, if the aim of your sample is to show an example for a specific platform running a specific kernel, i'm ok.
But if this example has to be maintained and integrated on some other platforms, i would suggest to change it...

wjliang commented on samples/subsys/ipc/openamp/remote_echo/README.rst:
```diff
@@ -0,0 +1,142 @@
+    CONFIG_HAVE_IMX_RPMSG=y
+    CONFIG_RPMSG=y
+    CONFIG_RPMSG_VIRTIO=y
+    CONFIG_IMX_RPMSG_TTY=m
```
Is there any reason to enable IMX_RPMST_TTY driver?
Just check the .c file, it looks like it implements the echo, from meta-openamp, there is userspace demo which uses the kernel rpmsg_char driver, as @arnop2 mentioned, there is also kernel sample driver.
Here is the OpenAMP apps for the kernel sample driver: https://github.com/OpenAMP/open-amp/blob/master/apps/examples/rpmsg_sample_echo/rpmsg-sample-ping.c

It looks like only the virtio operation implementation is different, maybe you can reuse that application.

wjliang commented on samples/subsys/ipc/openamp/remote_echo/src/main.c:
```diff
@@ -0,0 +1,299 @@
+	}
+
+	/* setup vdev */
+	vq[0] = virtqueue_allocate(VRING_SIZE);
```
this function will do dynamic memory allocation. Just in case, if you want to avoid dynamic allocation, you can statically defined the virtqueue. As your demo is RPMSG_REMOTE, you can try static virtqueue vq[2].

wjliang commented on samples/subsys/ipc/openamp/remote_echo/src/main.c:
```diff
@@ -0,0 +1,299 @@
+
+	/* Send data back to master */
+	memset(payload, 0, RPMSG_BUFFER_SIZE);
+	memcpy(payload, data, len);
```
You don't need to do memset 0.

wjliang commented:
Just from this commit: https://github.com/zephyrproject-rtos/zephyr/pull/11934/commits/62c0d7440ae55517df79ddb44bef6e29da46ba10
It looks like it is also doing echoing, which is similar to this pull request https://github.com/zephyrproject-rtos/zephyr/pull/11355  (@galak). Only the notification and the shared memory is different. just wonder if it can combine these two into one demo, and use configuration options to enable different configuration?

dbkinder reviewed (approved):
+1 for doc changes, thanks.

ioannisg commented:
@diegosueiro could you rebase this PR, if there's a plan to work on it again?

galak commented:
closing at this has be stale for sometime, feel free to re-open and update.
